### PR TITLE
Storage Emulator support in rules-unit-testing

### PR DESCRIPTION
**Pull request by samtstern**
### Discussion

Adds support for Storage emulator to the rules-unit-testing library.

TODO:

- [x] Add tests
- [x] Merge and release https://github.com/firebase/firebase-tools/pull/3332

Nice to have:

- [ ] Merge and release https://github.com/firebase/firebase-tools/pull/3337 (remove need for env)
- [ ] Merge and release https://github.com/firebase/firebase-tools/pull/3338 (make logs cleaner)

### Testing

See unit tests. Currently we have the following failures, so we have to skip a test in CI:

```
  1) Testing Module Tests
       initializeAdminApp() has admin access to storage:
     Error: Could not load the default credentials. Browse to https://cloud.google.com/docs/authentication/getting-started for more information.
      at GoogleAuth.getApplicationDefaultAsync (/home/runner/work/firebase-js-sdk/firebase-js-sdk/node_modules/google-auth-library/build/src/auth/googleauth.js:173:19
```

(1) is a known issue with Storage not being able to work without credentials even when using the emulator. @abeisgoat or @hiranya911 do you know of any way to fix this without adding service account credentials to the JS SDK test environment?

### API Changes

`initializeAdminApp()` has always been returning a `firebase.app.App` type from the client SDK but that has not been correct for a few releases! So I did change the type which is an API change, but it's to fix a bug.

### CODE REVIEW

**Comment by IchordeDionysos** [comment]
Very nice 😍
Thanks Sam

File: packages/rules-unit-testing/src/api/index.ts [hunk 1]
```diff
@@ -31,11 +31,16 @@ const DATABASE_ADDRESS_ENV: string = 'FIREBASE_DATABASE_EMULATOR_HOST';
...
 const DATABASE_ADDRESS_DEFAULT: string = 'localhost:9000';
 
-/** If any of environment variable is set, use it for the Firestore emulator. */
+/** If this environment variable is set, use it for the Firestore emulator. */
 const FIRESTORE_ADDRESS_ENV: string = 'FIRESTORE_EMULATOR_HOST';
 /** The default address for the local Firestore emulator. */
 const FIRESTORE_ADDRESS_DEFAULT: string = 'localhost:8080';
 
+/** If this environment variable is set, use it for the Storage emulator. */
+const STORAGE_ADDRESS_ENV: string = 'STORAGE_EMULATOR_HOST';
```
Comment (abeisgoat) [inline packages/rules-unit-testing/src/api/index.ts:40]: This is the cloud variable which requires the protocol to be specified like `STORAGE_EMULATOR_HOST=http://localhost:9199` we could use the `FIREBASE_STORAGE_EMULATOR_HOST=localhost:9199` however this only works in new releases of the admin SDK

**Review by abeisgoat — APPROVED** [review]

File: packages/rules-unit-testing/src/api/index.ts [hunk 2]
```diff
@@ -201,19 +216,29 @@ export function initializeTestApp(options: AppOptions): firebase.app.App {
...
 }
 
 export type AdminAppOptions = {
   databaseName?: string;
   projectId?: string;
+  storageBucket?: string;
 };
 /** Construct an App authenticated as an admin user. */
-export function initializeAdminApp(options: AdminAppOptions): firebase.app.App {
+export function initializeAdminApp(options: AdminAppOptions): app.App {
```
Comment (samtstern) [inline packages/rules-unit-testing/src/api/index.ts:233]: This type has been wrong for a while!

File: packages/rules-unit-testing/src/api/index.ts [hunk 3]
```diff
@@ -417,17 +491,22 @@ function initializeApp(
...
     app.database().goOffline();
     app.database().goOnline();
   }
   if (projectId) {
-    app.firestore().settings({
-      host: getFirestoreHost(),
-      ssl: false
-    });
+    const { hostname, port } = parseHost(getFirestoreHost());
+    app.firestore().useEmulator(hostname, port);
```
Comment (samtstern) [inline packages/rules-unit-testing/src/api/index.ts:505]: Moving to consistently use `useEmulator` across the board.

File: packages/rules-unit-testing/test/database.test.ts [hunk 4]
```diff
@@ -216,7 +220,24 @@ describe('Testing Module Tests', function () {
     });
   });
 
-  it('initializeAdminApp() has admin access', async function () {
+  it('initializeAdminApp() has admin access to RTDB', async function () {
```
Comment (samtstern) [inline packages/rules-unit-testing/test/database.test.ts:223]: Separated this into three tests so we can tell failures apart.

File: packages/rules-unit-testing/package.json [hunk 5]
```diff
@@ -15,7 +15,7 @@
     "build:deps": "lerna run --scope @firebase/rules-unit-testing --include-dependencies build",
     "dev": "rollup -c -w",
     "test:nyc": "TS_NODE_CACHE=NO TS_NODE_COMPILER_OPTIONS='{\"module\":\"commonjs\"}' nyc --reporter lcovonly -- mocha 'test/{,!(browser)/**/}*.test.ts' --config ../../config/mocharc.node.js",
-    "test": "firebase --project=foo --debug emulators:exec 'yarn test:nyc'",
+    "test": "FIREBASE_CLI_PREVIEWS=storageemulator STORAGE_EMULATOR_HOST=http://localhost:9199 firebase --project=foo --debug emulators:exec 'yarn test:nyc'",
```
Comment (samtstern) [inline packages/rules-unit-testing/package.json:18]: I had to add the `STORAGE_EMULATOR_HOST` variable to compensate for https://github.com/firebase/firebase-tools/pull/3337

**Review by yuchenshi — APPROVED** [review]

**Review by hsubox76 — APPROVED** [review]

**Review by sam-gc — APPROVED** [review]

File: packages/rules-unit-testing/package.json (hunk 5 above)
Comment (yuchenshi) [reply packages/rules-unit-testing/package.json:18]:
```suggestion
    "test": "FIREBASE_CLI_PREVIEWS=storageemulator firebase --project=foo --debug emulators:exec 'yarn test:nyc'",
```

Gotta clean this up now we've released the CLI with the fix

**Pull request merged**